Drupal 8: Batch API demo with Guzzle and Crawler usage examples.

mxr576 opened the pull request.

mxr576 committed:
Drupal 8: Batch API demo with Guzzle and Crawler usage examples.

mxr576 committed:
Typo.

mxr576 committed:
README.md added.

boobaa commented on d8/d8batch/README.md:
```diff
@@ -0,0 +1,11 @@
+Drupal's [Batch API](https://api.drupal.org/api/drupal/core!includes!form.inc/group/batch/8) did not change anything in Drupal 8, that is why this demo module also contains other useful examples from the Drupal 8 new features.
+
+#### About this module
+This module basically a demonstration of how can you import feed sources from a CSV file to the Drupal. The imported feed sources will be stored in nodes.
```
This module is…

boobaa commented on d8/d8batch/config/install/field.storage.node.field_feed_source_url.yml:
```diff
@@ -0,0 +1,19 @@
+uuid: 6d09794e-a019-42ae-9a43-67df9009c074
+langcode: en
```
Why does a configuration (entity?) need a language key if it doesn't have a label (or anything displayed to users)?

mxr576 replied:
Also a good question, my answer is the same as above, these file also came from Drupal CMI UI directly. But I'll try to investigate the reason.

![screen shot 2015-11-17 at 07 44 14](https://cloud.githubusercontent.com/assets/1755573/11204429/1480096c-8cff-11e5-95fc-4e498e2b2d4c.png)

mxr576 replied:
Core also contains language spec. settings in field storage conf. files.
http://cgit.drupalcode.org/drupal/tree/core/profiles/standard/config/install/field.storage.node.field_tags.yml?id=4f9265c

boobaa commented on d8/d8batch/README.md:
```diff
@@ -0,0 +1,11 @@
+#### About this module
+This module basically a demonstration of how can you import feed sources from a CSV file to the Drupal. The imported feed sources will be stored in nodes.
+
+The module also holds the definition of the necessary custom node type (Feed source) and its link field (field_feed_source_link) thanks to the new configuration management system.
```
You should provide detailed description of each such feature. Eg. what are the parts needed if I only want to have a custom content type provided by my module?

boobaa commented on d8/d8batch/config/install/field.field.node.feed_source.field_feed_source_url.yml:
```diff
@@ -0,0 +1,23 @@
+default_value: {  }
+default_value_callback: ''
+settings:
+  link_type: 16
```
What is link_type:16?

mxr576 replied:
Some magic number from the link.module, I assume. :)

boobaa commented on d8/d8batch/README.md:
```diff
@@ -0,0 +1,11 @@
+
+The module also holds the definition of the necessary custom node type (Feed source) and its link field (field_feed_source_link) thanks to the new configuration management system.
+
+The imported feed item links will be stored in the mentioned link field and the title of the generated nodes will be set to the feed source's original title with the help of the [Guzzle](http://docs.guzzlephp.org/en/latest/) and the [Crawler](http://symfony.com/doc/current/components/dom_crawler.html) libraries, which also part of the Drupal 8 (thanks to the Symfony).
```
…which are also…
…thanks to Symfony…

boobaa commented on d8/d8batch/config/install/field.field.node.feed_source.field_feed_source_url.yml:
```diff
@@ -0,0 +1,23 @@
+label: URL
+description: ''
+required: true
+translatable: false
```
If this is not translatable, then why does it need a langcode?

mxr576 replied:
Good question, all of these configuration files were exported with the Drupal CMI UI. I didn't wrote them manually.

boobaa commented on d8/d8batch/config/install/node.type.feed_source.yml:
```diff
@@ -0,0 +1,17 @@
+status: true
+dependencies:
+  module:
+    - menu_ui
```
Why should a content type depend on the menu_ui module?

mxr576 replied:
Probably, because the entity (content type) contains menu_ui settings. See below. I'll check what will happen if I remove these from the automatically generated file.

boobaa commented on d8/d8batch/d8batch.install:
```diff
@@ -0,0 +1,31 @@
+ * Implements hook_install().
+ */
+function d8batch_install() {
+  \Drupal\menu_link_content\Entity\MenuLinkContent::create([
```
Isn't this available as a service?

boobaa commented on d8/d8batch/d8batch.install:
```diff
@@ -0,0 +1,31 @@
+ * Implements hook_uninstall().
+ */
+function d8batch_uninstall() {
+  $mids = \Drupal::entityQuery('menu_link_content')
```
Isn't this available as a service?

mxr576 replied:
Yes there is a service so I've replaced this solution with it. Just a short notice: When you create a new link with the service it can not be edited on the UI (ex.: title), but with this solution the links could be modified on the UI.

boobaa commented on d8/d8batch/d8batch.module:
```diff
@@ -0,0 +1,161 @@
+    $url = Url::fromUri($url);
+    if (!d8batch_check_if_feed_source_already_exists($url)) {
+      try {
+        $client = new Client();
```
I _think_ it would look cooler if the Client() was dependency-injected via the $context. I _think_, but I don't know if it's possible at all.

boobaa commented on d8/d8batch/d8batch.module:
```diff
@@ -0,0 +1,161 @@
+ */
+function _d8batch_batch_finished($success, $results, $operations) {
+  if ($success) {
+    $message = \Drupal::translation()
```
Isn't this available as a service?

mxr576 replied:
I did not find any info about this. The official documentation uses the same solution:
https://api.drupal.org/api/drupal/core!includes!form.inc/group/batch/8

boobaa commented on d8/d8batch/d8batch.module:
```diff
@@ -0,0 +1,161 @@
+ *   TRUE if node exists, else FALSE.
+ */
+function d8batch_check_if_feed_source_already_exists(Url $url) {
+  $nids = \Drupal::entityQuery('node')
```
Isn't this available as a service?

boobaa commented on d8/d8batch/d8batch_demo_feed_sources.csv:
```diff
@@ -0,0 +1,378 @@
+http://www.bdlive.co.za/?service=rss
+http://www.theglobeandmail.com/?service=rss
+http://feeds.bbci.co.uk/news/rss.xml#
+http://sputniknews.com/export/rss2/index.xml
```
Why having this huge list? I _think_ a smaller number would be quite as sufficient.

mxr576 replied:
I reduced the number of the input.

boobaa commented on d8/d8batch/src/Form/D8BatchProcessForm.php:
```diff
@@ -0,0 +1,74 @@
+   * {@inheritdoc}.
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $form['actions'] = [
```
It would be wise to display the path where you are importing that list from.

boobaa commented on d8/d8batch/README.md:
```diff
@@ -0,0 +1,11 @@
+### D8 Batch
+
+#### tl;dr;
+Drupal's [Batch API](https://api.drupal.org/api/drupal/core!includes!form.inc/group/batch/8) did not change anything in Drupal 8, that is why this demo module also contains other useful examples from the Drupal 8 new features.
```
Am I correct that those useful examples are:
- using Guzzle (to fetch something from an URL),
- using Crawler (to walk the XML);
- providing a custom content type?

What happens if a different module already provides the same content type?

mxr576 replied:
I've moved the configuration files from the config/install to the config/optional, this way this conflict will be ignored. However I'm still not sure what would be the proper solution for storing node type/field configuration in modules.

mxr576 committed:
Various changes.

mxr576 committed:
Revert if statement changes.

mxr576 committed:
UUID-s were removed from the configs.

mxr576 committed:
- Move Feed Source node type definition to separate module.
- Make sure that the Feed Source nodes will be removed too after the Feed
Source node type definition removed from the active configuration.
- Other improvements.

mxr576 committed:
README.md updated.

mxr576 committed:
Form alter example in D8, which add an extra "Remove all XY nodes" button to
the node type delete forms, if there is an existing node in the Drupal with
the given node type.
Generally, it is impossible to remove a node type (or any other entity type)
configuration from the active configuration, while at least one entity
exists in the system with the given (entity) type.

boobaa commented on d8/d8batch/d8batch.module:
```diff
@@ -0,0 +1,170 @@
+    $url = Url::fromUri($url);
+    if (!d8batch_check_if_feed_source_already_exists($url)) {
+      try {
+        $client = new \GuzzleHttp\Client();
```
This could look better, I guess: `use \GuzzleHttp\Client as GuzzleClient;` (`as …` is only needed if there's a clash on the `Client` part).

boobaa commented on d8/d8batch/d8batch.module:
```diff
@@ -0,0 +1,170 @@
+          // If title can not be extracted, then use the URL as title.
+          if (!($title = _d8batch_get_title_of_feed_source($response->getBody()
+            ->getContents()))
+          ) {
```
This syntax looks really weird to me. Can you point me to an example in core which uses this same syntax?

mxr576 replied:
This is not intended to be like this, I'll fix it.

boobaa commented on d8/d8batch/d8batch.module:
```diff
@@ -0,0 +1,170 @@
+            'title' => $title,
+            'field_feed_source_url' => [
+              'uri' => $url->getUri(),
+            ],
```
It's worth mentioning (in the readme) that you don't have to write a miles-long stuff here like in D7's `$node->field_foo[LANGUAGE_NONE][0]['value']`.

boobaa commented on d8/d8batch/d8batch.module:
```diff
@@ -0,0 +1,170 @@
+    }
+  }
+  catch (InvalidArgumentException $e) {
+    // Feed source url not valid.
```
…is not valid…

boobaa commented on d8/d8batch/d8batch.module:
```diff
@@ -0,0 +1,170 @@
+function _d8batch_batch_finished($success, $results, $operations) {
+  if ($success) {
+    if (empty($results['success']) && empty($results['failed'])) {
+      drupal_set_message(t('All feed sources successfully imported before!'));
```
…sources were successfully…

boobaa commented on d8/d8batch/modules/d8feedsourcect/config/install/core.entity_view_display.node.feed_source.default.yml:
```diff
@@ -0,0 +1,32 @@
+    label: inline
+    settings:
+      trim_length: 80
+      url_only: false
```
If you're storing only the URLs, why do you want to display anything else?

boobaa commented on d8/d8batch/modules/d8feedsourcect/config/install/field.field.node.feed_source.field_feed_source_url.yml:
```diff
@@ -0,0 +1,25 @@
+default_value: {  }
+default_value_callback: ''
+settings:
+  link_type: 16
```
I know it's not your fault but core's, but what does 16 mean in this context? :)

boobaa commented on d8/d8batch/modules/d8feedsourcect/config/install/field.storage.node.field_feed_source_url.yml:
```diff
@@ -0,0 +1,21 @@
+module: link
+locked: false
+cardinality: 1
+translatable: true
```
I don't really get what is translatable and what is not. To me it looks like the same stuff is both translatable and untranslatable. What's the story here?

boobaa commented on d8/d8batch/modules/d8feedsourcect/d8feedsourcect.info.yml:
```diff
@@ -0,0 +1,10 @@
+  - field
+  - link
+  - node
+  - path
```
Why depending on path.module? I thought these nodes have only a link field, but not a (custom) path.

boobaa commented on d8/d8batch/modules/d8feedsourcect/d8feedsourcect.module:
```diff
@@ -0,0 +1,54 @@
+use Drupal\Core\Entity\EntityStorageException;
+
+/**
+ * Implements callback_batch_operation().
```
I don't think there's such a hook, so this should say `Batch operation callback for …`.

boobaa commented on d8/d8batch/modules/d8feedsourcect/d8feedsourcect.module:
```diff
@@ -0,0 +1,54 @@
+  $current_nids = array_slice($context['sandbox']['nids'], $context['sandbox']['progress'], 10);
+
+  foreach ($current_nids as $nid) {
+    if ($node = \Drupal\node\Entity\Node::load($nid)) {
```
`use …` this class, pls. ;)

boobaa commented on d8/d8batch/modules/d8feedsourcect/d8feedsourcect.module:
```diff
@@ -0,0 +1,54 @@
+}
+
+/**
+ * Implements callback_batch_finished().
```
Again: I don't think it's a hook implementation (see above).

boobaa commented on d8/d8batch/modules/d8feedsourcect/d8feedsourcect.module:
```diff
@@ -0,0 +1,54 @@
+ */
+function _d8feedsourcect_delete_feed_source_node_batch_finished($success, $results, $operations) {
+  if ($success && !$results['error']) {
+    drupal_set_message(t('All feed source nodes successfully deleted!'));
```
…nodes have been successfully…

boobaa commented on d8/d8batch/src/Form/D8BatchProcessForm.php:
```diff
@@ -0,0 +1,74 @@
+    $batch = [
+      'title' => t('Importing feed sources'),
+      'operations' => [],
+      'progress_message' => t('Processed @current feed source from @total.'),
```
I'm not 100% sure these messages should be enclosed in a `t()` call. I _think_ these are/should be run through the translation service when they're displayed, not when they're set up.

mxr576 replied:
We should asks Gabor Hojtsy about this. :)

boobaa replied:
On it – dawehner is already answering my questions on IRC.

boobaa commented on d8/d8forms/d8forms.module:
```diff
@@ -0,0 +1,139 @@
+ * Add ability to node type delete forms to remove existing nodes with node
+ * types.
+ */
+function d8forms_form_node_type_delete_form_alter(&$form, \Drupal\Core\Form\FormStateInterface $form_state, $form_id) {
```
`use …` this class, pls. ;)

boobaa commented on d8/d8forms/d8forms.module:
```diff
@@ -0,0 +1,139 @@
+function d8forms_form_node_type_delete_form_removes_nodes_submit(&$form, \Drupal\Core\Form\FormStateInterface $form_state) {
+  $build_info = $form_state->getBuildInfo();
+  $nids = d8forms_get_node_type_nodes($build_info['callback_object']->getEntity()
+    ->get('type'));
```
This syntax looks weird to me.

boobaa commented on d8/d8forms/d8forms.module:
```diff
@@ -0,0 +1,139 @@
+          [$nids, $build_info['callback_object']->getEntity()->get('name')],
+        ],
+      ],
+      'progress_message' => '',
```
Is this really needed?

mxr576 replied:
Yes, because I would like to hide the original (miss-leading) progress message. My module displays better info than that about the status of the current progress. But this could be simple just NULL, thanks for noticing that.

boobaa commented on d8/d8forms/d8forms.module:
```diff
@@ -0,0 +1,139 @@
+  $current_nids = array_slice($context['sandbox']['nids'], $context['sandbox']['progress'], 10);
+
+  foreach ($current_nids as $nid) {
+    if ($node = \Drupal\node\Entity\Node::load($nid)) {
```
`use …` this class, pls. :)

boobaa commented on d8/d8forms/d8forms.module:
```diff
@@ -0,0 +1,139 @@
+}
+
+/**
+ * Implements callback_batch_operation().
```
I don't think it's a hook implementation (see above).

mxr576 replied:
Yes, there is: https://api.drupal.org/api/drupal/modules!system!form.api.php/function/callback_batch_operation/7

boobaa commented on d8/d8forms/d8forms.module:
```diff
@@ -0,0 +1,139 @@
+}
+
+/**
+ * Implements callback_batch_finished().
```
I don't think it's a hook implementation (see above).

mxr576 replied:
https://api.drupal.org/api/drupal/modules!system!form.api.php/function/callback_batch_finished/7

boobaa commented on d8/d8forms/d8forms.module:
```diff
@@ -0,0 +1,139 @@
+
+use Drupal\Core\Routing\RouteMatchInterface;
+
+define('D8FORMS_MODULE_NAME', 'd8forms');
```
Hmm, this is the first time I'm seeing the module's machine name encoded this way. I don't really understand the reasoning behind this.

mxr576 replied:
Because I need the module's name in various places (for example in watchdog()) I thought this could be a good idea the prevent typos in my code. I've never used this "pattern" before either.

mxr576 committed:
Minor modifications and code cleanup.

mxr576 committed:
README.md updated.

mxr576 committed:
Update README.md

mxr576 committed:
Update README.md
